### ISSUE TITLE
FP8 blockwise scaling backward produces near-zero dGrad when loading pre-quantized weights (DeepSeek V4 Flash)

### ISSUE THREAD

**Issue by HorizonChaser** [modelscope/mcore-bridge#125]
### Checklist / 检查清单

- [x] I have searched existing issues, and this is a new bug report. / 我已经搜索过现有的 issues，确认这是一个新的 bug report。

### Bug Description / Bug 描述

When using `fp8_param_gather=True` with FP8 pre-quantized HuggingFace checkpoints (for example I used  DeepSeek-V4-Flash from modelscope), the backward pass dGrad GEMM reads stale **all-zero** `_columnwise_data`, producing incorrect near-zero gradients.

## Root Cause

`GPTBridge._set_param()` updates `_rowwise_data` from the HF checkpoint but does not regenerate `_columnwise_data`, which is consumed by the backward dGrad GEMM.

### Step 1: TE `reset_parameters()` quantizes all-zero params

During model initialization, TE quantizes parameters **before** checkpoint loading. At this point all params are zero, so both `_rowwise_data` and `_columnwise_data` are allocated with all-zero content:

> [`transformer_engine/pytorch/module/base.py#L1459-L1462`](https://github.com/NVIDIA/TransformerEngine/blob/v2.15.0/transformer_engine/pytorch/module/base.py#L1459-L1462)
> ```python
> quantizer.set_usage(rowwise=True, columnwise=torch.is_grad_enabled())
> quantizer.internal = False
> # ...
> param = quantizer(param)  # param is all zeros → columnwise_data = all zeros
> ```

### Step 2: `_set_param()` only updates rowwise data

mcore-bridge loads HF FP8 weights but only writes `_rowwise_data` and `_rowwise_scale_inv`, leaving `_columnwise_data` stale:

> [`src/mcore_bridge/bridge/gpt_bridge.py#L214-L218`](https://github.com/modelscope/mcore-bridge/blob/9b5ac5b/src/mcore_bridge/bridge/gpt_bridge.py#L214-L218)
> ```python
> else:
>     tensor = tensor.view(torch.uint8)
>     param._rowwise_data.data.copy_(tensor)       # updated
>     self._copy_scale_inv(param, hf_scale_inv)    # updated (rowwise only)
>     del param.get_high_precision_init_val
>     # _columnwise_data is NOT updated or invalidated
> ```

### Step 3: `update_usage()` skips regeneration

In training, `update_usage(columnwise_usage=True)` is called before the backward GEMM. It checks whether `_columnwise_data` needs to be created — but since the buffer already exists (allocated in Step 1, just all zeros), it skips `_create_columnwise()`:

> [`transformer_engine/pytorch/tensor/storage/float8_blockwise_tensor_storage.py#L454-L455`](https://github.com/NVIDIA/TransformerEngine/blob/v2.15.0/transformer_engine/pytorch/tensor/storage/float8_blockwise_tensor_storage.py#L454-L455)
> ```python
> if self._columnwise_data is None or self._columnwise_scale_inv is None:
>     self._create_columnwise()
> # ↑ condition is False — _columnwise_data exists (all zeros), so skipped
> ```

### Step 4: cuBLAS reads zero columnwise data → dGrad ≈ 0

`general_gemm(layout="NN")` passes `_columnwise_data` directly to cuBLAS for the weight matrix:

> [`transformer_engine/common/gemm/cublaslt_gemm.cu` (CanonicalizeGemmInput)](https://github.com/NVIDIA/TransformerEngine/blob/v2.15.0/transformer_engine/common/gemm/cublaslt_gemm.cu)
> ```cpp
> if (!is_A_transposed) {
>     ret.A = A.columnwise_data.dptr;          // ← all zeros
>     ret.A_scale_inv = A.columnwise_scale_inv.dptr;
> }
> ```
>
> Result: `dX = dY @ 0 = 0` for all affected layers.

## Result

- Gradient norm drops from ~150 (BF16 baseline is 154) to ~57 on step 1
- Training appears to converge but produces suboptimal weights

## Environment

- Transformer Engine 2.15.0, CUDA 12.6
- torch 2.12.0+cu126
- ms-swift 4.4.0.dev0 (commit 46f4df6)
- mcore-bridge (latest main, commit 9b5ac5b)
- DeepSeek-V4-Flash (236B, LoRA finetune)
- 8 x NVIDIA H20

## Script Used for lora Finetuning

```bash
#!/bin/bash
source /root/.venv/bin/activate

# Force NCCL to use compat driver (575) instead of kernel driver (535)
export LD_PRELOAD=/usr/local/cuda/compat/libcuda.so.1

export PYTORCH_CUDA_ALLOC_CONF='expandable_segments:True'
export NPROC_PER_NODE=4
export NNODES=2
export MASTER_ADDR=[IP_ADDR]
export MASTER_PORT=29500
export CUDA_VISIBLE_DEVICES=0,1,2,3

# Set NODE_RANK from environment or default to 0
export NODE_RANK=${NODE_RANK:-0}

export NCCL_DEBUG=WARN
export NCCL_IB_DISABLE=1
export NCCL_SOCKET_IFNAME=eth1

megatron sft \
    --model '/root/models/DeepSeek-V4-Flash/'\
    --save_safetensors true \
    --dataset '[DATASET]' \
    --merge_lora false \
    --load_from_cache_file true \
    --add_non_thinking_prefix true \
    --loss_scale ignore_empty_think \
    --split_dataset_ratio 0.05 \
    --tuner_type lora \
    --target_modules linear_qkv linear_proj linear_fc1 linear_fc2 \
    --lora_rank 16 \
    --lora_alpha 32 \
    --fp8_recipe blockwise \
    --fp8_format e4m3 \
    --fp8_param_gather true \
    --tensor_model_parallel_size 1 \
    --expert_model_parallel_size 8 \
    --micro_batch_size 1 \
    --global_batch_size 16 \
    --padding_free false \
    --group_by_length true \
    --recompute_granularity full \
    --recompute_method uniform \
    --recompute_num_layers 2\
    --mhc_recompute_layer_num 1 \
    --moe_permute_fusion true \
    --moe_grouped_gemm true \
    --moe_shared_expert_overlap true \
    --moe_aux_loss_coeff 1e-3 \
    --finetune true \
    --cross_entropy_loss_fusion true \
    --optimizer_cpu_offload true \
    --main_grads_dtype bf16 \
    --lr 5e-5 \
    --lr_warmup_fraction 0.1 \
    --min_lr 1e-6 \
    --clip_grad 1.0 \
    --output_dir megatron_output/DeepSeek-V4-Flash \
    --max_length 1024 \
    --num_train_epochs 1 \
    --dataloader_num_workers 8 \
    --dataset_num_proc 8 \
    --truncation_strategy right \
    --no_save_optim true \
    --no_save_rng true \
    --logging_steps 1 \
    --use_precision_aware_optimizer true \
    --sequence_parallel true \
    --attention_backend flash
```

### Additional Information / 补充信息

Training in FP8 with original mcore-bridge:

<img width="1390" height="602" alt="Image" src="https://github.com/user-attachments/assets/345c02c2-c738-4f13-80f3-9d368066faea" />

Training in FP8 with our fix:

<img width="1388" height="427" alt="Image" src="https://github.com/user-attachments/assets/8c0f0212-bce2-4c25-8f30-c8b04e056496" />

Training in BF16 with original mcore-bridge:

<img width="1389" height="597" alt="Image" src="https://github.com/user-attachments/assets/4e7de075-bcaa-4d8b-b264-6566356dfda0" />

**Comment by HorizonChaser** [comment]
I've come up with a fix in #126 , please take a look :)

**Comment by Jintao-Huang** [comment]
thanks!
